### ISSUE TITLE
Moving images: implementation plan

### ISSUE THREAD

**Issue by bernhold** [betterscientificsoftware/bssw.io#738]
I propose this to supersede #703.

Parallactic would prefer an approach which would allow a smoother stepwise transition rather than a hard cut-over from the current approach to the new one.  This should make it easier to ensure continuous operation of the site.

## Tasks

- [x] We copy all of the contents of https://github.com/betterscientificsoftware/images into https://github.com/betterscientificsoftware/bssw.io/images on master
- [x] We add a relative image reference to the "kitchen sink" test article (https://github.com/betterscientificsoftware/bssw.io/blob/master/CuratedContent/kitchen-sink-TEST.md)  This is *not* expected to work yet on the front-end, but should work with authoring tools.
- [x] Write scripts that will do the translation from the separate 'images' repo into the 'bssw.io' repo and update all of the references in the `*.md` documents (see PR #752)
- [x] Change images/README.md to reflect new process (see PR #754)
- [x] Create at least one new PR with new content with images and test the updated process. and verify ... **See #794**
- [x] Final review the plan to do the translation ... **Completed at 5/13/2021 meeting**
- [x] Parallactic deploys transitional code which will leave fully qualified URLs as they are, while mapping relative URLs referencing the images directory appropriately.  The transitional code also deals with some corner cases that appear in Site/Homepage.md and in the Fellows area.  At this point, the image in the kitchen sink article should work as intended.
- [x] Do the translation of everything on 'master' (**scheduled for 5/13/2021**): 
  - [x] Create local branch.
  - [x] Copy anything added to images since the initial copy into the local branch.
  - [x] Run image reference translation scripts on all `*.md` files.
  - [x] Create PR for local branch and get merged to 'preview' branch. ... **See PR #909**
  - [x] Verify proper front-end translation in 'preview' branch (should reference the images directory on the *preview* branch).
  - [x] Merge the PR.
  - [x] Verify proper front-end translation in production site (should reference the images directory on the *master* branch)
- [x] At this point, the separate images repo should be considered read-only.  All changes and new images should be done in the bssw.io/images/ directory. 
- [x] Parallactic changes sponsor logo references for https://bssw.io/pages/meet-our-fellows
- [x] Archive https://github.com/betterscientificsoftware/images repo.
- [x] Create a PR with updated EB docs, PR checklist, remove transition note from top of `images/README.md` etc. for the updated handling of 'images' (and have it reviewed) ... **See PR #1027**
- [x] Update imgs in outstanding PRs to reference local bssw.io/images/ directory.
- [x] Manually translate all contents of https://github.com/betterscientificsoftware/bssw.io/blob/master/Site to produce a PR.  (Many files in this tree use a different syntax to refer to images, so we need to do a careful check and make sure we get all of them.  The references are in the form "/images/..." but are *not* necessarily in image tags -- they're treated specially by the front-end.  Given our approach, references starting at the root of the repo may just continue to work as-is.  But we need to check carefully on the preview site.) **@bernhold thinks this has already been taken care of as of 2021-06-18.  Matt has scanned the entire site and all but two images were referencing the right repo, and those two have now been fixed.**
- [x] Parallactic deploys final code, which will allow us to create subdirectories under images if we want to (for future (re)organization of the images space.
- [x] Reverify expected image mappings by front-end.
- [x] Mark as DONE!

At the completion of this work, follow-on activities might include:

* Removal of unused images
* Using subdirectories in the images folder to better organize images (e.g. separating fellowship, blog, and other images)
  - @bernhold: this requires further changes to the front-end to support.  They're on PC's to-do list, but we need to verify that they're complete before we start using subdirectories.  Also, I believe that filenames will need to remain globally unique. 
* Renaming images to follow new file naming conventions (yet to be defined)

**Comment by bartlettroscoe** [comment]
> I propose this to supersede #703.

@bernhold, why?

**Comment by bartlettroscoe** [comment]
> We copy all of the contents of https://github.com/betterscientificsoftware/images into https://github.com/betterscientificsoftware/bssw.io/images on master

But then you have the problem of people modifying images in both places, no?

**Comment by bernhold** [comment]
As soon as Parallactic puts the transitional code in there, the images directory will work and all new/changed images should go there.  That should happen quickly.  We should be to consider the separate images repo as effectively read-only now, and really read-only as soon as the transitional code is in.  In fact, I will modify the checklist to reflect that.

**Comment by bartlettroscoe** [comment]
@bernhold, I can see that you copied a bunch images into the `bssw.io/images/` directory and pushed to 'master' in the commit f51c5c7ed57c35714a84cd38f63e2e04f7493ce6 with log message:

```
Copying in contents of images repo
    
Per #703 and #738, we're copying the contents of the [images repository](https://github.com/betterscientificsoftware/images) into the bssw.io repository, with the eventual goal of eliminating the separate images repo.
```

Did you just copy the entire contents of the separate `images` repo?  If not, then how did you select the subset to copy over?

Also, one small critique; generally one wants to write commit messages that wrap the body text.  Recommendations are for 72 chars wide but a bit wider is also readable in most contexts.  See, for example:

* https://chris.beams.io/posts/git-commit/
* https://www.freecodecamp.org/news/writing-good-commit-messages-a-practical-guide/
* https://medium.com/@steveamaza/how-to-write-a-proper-git-commit-message-e028865e5791
* https://gist.github.com/turbo/efb8d57c145e00dc38907f9526b60f17
* https://dev.to/helderburato/patterns-for-writing-better-git-commit-messages-4ba0

(All of those recommend 72 chars wrapping.  Not sure where this 72 char width first came from but there it is.)

**Comment by bernhold** [comment]
> Did you just copy the entire contents of the separate `images` repo? 

Yes.  We can clean it up and reorganize it later. If we want to.

> (All of those recommend 72 chars wrapping. Not sure where this 72 char width first came from but there it is.)

Presumably that's from the days when your git repository was a pile of punch cards :-)

**Comment by bartlettroscoe** [comment]
> Yes. We can clean it up and reorganize it later. If we want to.

But that is tool late.  Once you commit the files to 'master' (which you have), then they are in the git history **forever** and take up space on every clone of the repo.  The only way to really remove them is to filters the branches which creates a big mess of its own.

> Presumably that's from the days when your git repository was a pile of punch cards :-)

It problem is the formatting of the commit message on the command-line with `git log`.

The git standard for commit log message is pretty simple and can be expressed like [here](https://ideas-productivity.org/resources/howtos/git-tutorial-and-reference-collection/beginner-tips/) as:

```
Summary line for changes (50 chars is target but a little longer okay)

More detailed explanatory text in the body (wrapped to about 72 chars).
The blank line separating the summary from the body is critical to make
commands like 'log --oneline', 'rebase -i', and others readable and
useful.

If you use an issue tracker, put references to them like:

Resolves: #123
Also see: #456, #789
```

I break some of these guidelines (like summary longer than 50 chars and only mentioning issue trackers at the bottom).  But the point is that no tool will reformat the message body and it assumes that the commit author will format the body for things like bullets, etc.  It is always assumed to be raw text.

**Comment by bartlettroscoe** [comment]
@bernhold, @prwolfe 

PR #752 has the completed scripts to do the translation.  There is also a pointer in that PR to another branch and PR that shows what the converted files look like.

Can you please review PR #752 so it can be merged to 'master' and used later to convert the refs to the images?

NOTE: We can merge PR #752 at any time.  That is just the tools to do the conversion, not the actual conversion of the `*.md` files.

**Comment by bartlettroscoe** [comment]
@bernhold, @prwolfe 

FYI: PR #754 fixes the copied over `images/README.md` file.  Please review so it can be merged.

@bernhold, 

I have some questions about the updated process [above](https://github.com/betterscientificsoftware/bssw.io/issues/738#issue-832323849).  I added some notes and questions in **bold** with **bartlettroscoe**.

One thing I just noticed from the example:

* https://github.com/betterscientificsoftware/bssw.io/blob/master/CuratedContent/kitchen-sink-TEST.md

is that a reference from an `*.md` file to `/images/<file-name>` like:

```
<img src='/images/Blog_1119_WorkThankful.jpg' />
```

works from any `*.md` file from any subdirectory in the repo?  If that is true (and it seems to be true), then the translation scripts in PR #752 are unnecessary since you can just do a raw text replacement of `https://github.com/betterscientificsoftware/images/raw/master/` and `https://github.com/betterscientificsoftware/images/blob/master/` with `/images/` and you are done.

Will the web contractors implementation also work with absolute references `/images/<file-name>`?

**Comment by bartlettroscoe** [comment]
How embarrassing (for me).  He official page:

* https://guides.github.com/features/mastering-markdown/

shows the images example:

```
![GitHub Logo](/images/logo.png)
Format: ![Alt Text](url)
```

where it uses the absolute repository path `/images/logo.png`.  So this is what GitHub recommends (and that makes sense).

@bernhold, will the updated website translator respect absolute repository paths to images like  `/images/<file-name>`?

**Comment by bartlettroscoe** [comment]
Hum, it appears that all markdown parsers don't recognizes absolute repository paths like `/images/<file-name>` as per:

* https://github.com/shd101wyy/markdown-preview-enhanced/issues/34

so I am not crazy for not assuming that would work off the bat.

But this is not make clear in:

* https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-readmes

I can't seem to find a complete and authoritative description of the rules of relative and absolute links to fines in a github repo and in other markdown renderers.  For example, [this Microsoft markdown renderer](https://docs.microsoft.com/en-us/contribute/how-to-write-links) does not seem to support absolute repository references like `/images/<file-name>`.

Also, it is not clear what the rules are for repository references for github pages site (see [here](https://github.community/t/image-path-when-the-file-isnt-in-the-root/10499/7)).

So if someone were to use a local markdown renderer, it is not clear if an absolute repository reference like `/images/<file-name>` would work for them.  So it it safer to recommend to use relative path references like `../..//images/<file-name>` instead?  That seems to be supported by all markdown renderers (which makes sense since you don't need to know about the concept of a "repository" base directory for this to work).

**Comment by bartlettroscoe** [comment]
We discussed this in the BSSW EB meeting today and @bernhold agreed that it is better to use the relative path `../../images/<file-name>` instead of the absolute path `/images/<file-name>` because the latter does not work with some local Markdown editors like MS Visual Studio.  Therefore, we should keep the translation tool in PR  #750 as is and merge it.

**Comment by JeanShuler** [comment]
This is what I understood also. thanks

**Comment by bernhold** [comment]
@bartlettroscoe I made some revisions to the checklist up top in response to your comments.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe I made some revisions to the checklist up top in response to your comments.

@bernhold, thanks.  I updated the "Tasks" [above](https://github.com/betterscientificsoftware/bssw.io/issues/738#issue-832323849) again given your feedback (removing our comments) and the PRs just approved and merged.

**Comment by bernhold** [comment]
The transitional code for images is now in production.  Searching for "kitchen sink" on the preview site should return four test pages in different directories in the repo that exercise relative URLs.

And AFAIK, all of the images from the old repo have been copied over.  This should be independently confirmed before we check it off.

**Comment by bartlettroscoe** [comment]
> The transitional code for images is now in production

@bernhold, is there anything left blocking us to finish this?  Looking at the list [above](https://github.com/betterscientificsoftware/bssw.io/issues/738#issue-832323849), I think the rest is on us.  Should I just start knocking these off?

**Comment by bernhold** [comment]
@bartlettroscoe  yes, it is in our court now.  I haven't had time and it will probably be at least a week before I do.

**Comment by bartlettroscoe** [comment]
>  I haven't had time and it will probably be at least a week before I do.

@bernhold, do you want some help with this?  I think I can do almost all of this (except for some of the obscure bssw.io issues listed [above](https://github.com/betterscientificsoftware/bssw.io/issues/738#issue-832323849)).

**Comment by bartlettroscoe** [comment]
FYI: As discussed at the BSSW.io meeting today, I will do the translation on Monday 5/17/2021.  I will send out an email to the @betterscientificsoftware/bssw-editorial-board email list before and after I make that translation.  I would do it now except we need to wait for the publish of a blog article.

**Comment by bartlettroscoe** [comment]
Here is what I will do on Monday 5/17/2021:

* Send out an email to `bssw-editorial@<redacted>` stating when the translation will take place
* Do the translation in a local branch
* Post a PR for the translation
* Ask for a review of the 
* Look at the converted articles on preview.bssw.io
•	Merge the PR

**Comment by bartlettroscoe** [comment]
Closed by accident

**Comment by bartlettroscoe** [comment]
FYI: I updated the list of **Tasks** above for what I will actually do on Monday.  Specifically, I will do all of the items under:

* Do the translation of everything on 'master' (**scheduled for 5/13/2021**):

**Comment by bernhold** [comment]
@bartlettroscoe we found a bug in the transitional image handling.  We need to put the translation step on hold until it gets fixed.

**Comment by bartlettroscoe** [comment]
> we found a bug in the transitional image handling.

@bernhold, what bug and who is "we"?

I just ran it and it looks correct to me.  I posted the translated files in PR #858 that I just posted.  What files in that PR are not updated correctly?

The only links that were not replaced where:

```
$ find Articles -name "*.md" -exec grep -nH betterscientificsoftware/images {} \;
Articles/Blog/BlogArticleSkeletonA.md:7:The hero image must be this dimension ((1125 x 432 pixels); we can crop an image to the required size.  Put it in the images repo:  https://github.com/betterscientificsoftware/images
Articles/Blog/ReferencesInMarkdownHybridApproach-wikized.md:120:![](https://raw.githubusercontent.com/betterscientificsoftware/images/mcm86-19feb21-inplace-wikize-refs/wikize-refs-docs1.png)
Articles/Blog/ReferencesInMarkdownHybridApproach-wikized.md:125:![](https://raw.githubusercontent.com/betterscientificsoftware/images/mcm86-19feb21-inplace-wikize-refs/wikize-refs-docs2.png)
Articles/Blog/ReferencesInMarkdownHybridApproach-wikized.md:131:![](https://raw.githubusercontent.com/betterscientificsoftware/images/mcm86-19feb21-inplace-wikize-refs/wikize-refs-docs3.png)
Articles/Blog/ReferencesInMarkdownHybridApproach.md:120:![](https://raw.githubusercontent.com/betterscientificsoftware/images/mcm86-19feb21-inplace-wikize-refs/wikize-refs-docs1.png)
Articles/Blog/ReferencesInMarkdownHybridApproach.md:125:![](https://raw.githubusercontent.com/betterscientificsoftware/images/mcm86-19feb21-inplace-wikize-refs/wikize-refs-docs2.png)
Articles/Blog/ReferencesInMarkdownHybridApproach.md:131:![](https://raw.githubusercontent.com/betterscientificsoftware/images/mcm86-19feb21-inplace-wikize-refs/wikize-refs-docs3.png)
```

Those can be fixed manually by copying in those files to the bssw.io/images/ directory in the PR branch.

> We need to put the translation step on hold until it gets fixed.

What is the bug that would put this on hold?

**Comment by bernhold** [comment]
@bartlettroscoe "we" is the folks working on the current blog article and me.  The bug is that the front-end is not processing the hero references properly (they are not simple <img> tags in the published HTML, so any hero with a relative URL will just be blank at present.

**Comment by bartlettroscoe** [comment]
> The bug is that the front-end is not processing the hero references properly (they are not simple  tags in the published HTML, so any hero with a relative URL will just be blank at present.

@bernhold, okay, so this is a bug in the bssw.io site generator, not in the translation script I wrote.

That is a good reason to wait for the transition.  I will remove that from my calendar and wait to hear from you.

**Comment by bernhold** [comment]
> @bernhold, okay, so this is a bug in the bssw.io site generator, not in the translation script I wrote.

Ah, yes.  That was obvious (to me).  I'm not impugning your code! Sorry

**Comment by bartlettroscoe** [comment]
CC: @rinkug 

@bernhold, at the last BSSW.io meeting, @curfman reported that the issues of image handling on the bssw.io site have been resolved.  Do you concur?  If so, can I pull the trigger on this translation of all of the `*.md` files to point to local images in the local `bssw.io/images/` directory?  If so, I can schedule that for next Tuesday morning.

**Comment by bernhold** [comment]
@bartlettroscoe Yes, that's correct.  I was going to discuss it on tomorrow's operational call.  But yes, I think Tuesday would be good.  Thanks

**Comment by bartlettroscoe** [comment]
@bernhold and @prwolfe, I created PR #909 to do the basic translation of the `*.md` files.  But as I note in https://github.com/betterscientificsoftware/bssw.io/pull/909#issuecomment-854181333 it looks like the image rendering/translation code for the bssw.io site generator is not quite working yet.

**Comment by bernhold** [comment]
I've asked Matt Stone to run his spider to extract the image URLs, which should allow us to rigorously check for places that haven't been translated.

**Comment by bernhold** [comment]
Got the results back from Matt.  There are exactly two images that are referencing the old repo, the DOE and NSF logos on https://bssw.io/pages/meet-our-fellows.  That is not actually content that's under our control.  Clara (Parallactic) will fix it and let us know.

The scan also uncovered one image that's entirely outside of our space, the Sphinx logo at the bottom of https://bssw.io/items/an-introduction-to-sphinx.  Even though it is not strictly necessary, I'll change that too.

**Comment by bernhold** [comment]
Sphinx image changed in PR #924

**Comment by bartlettroscoe** [comment]
@bernhold and @prwolfe, can you please do a quick review of PR #1027?

@bernhold, note this PR also updates the instructions at [Front page carousel highlight slides](https://betterscientificsoftware.github.io/bssw.io/bssw_content_highlighting.html#front-page-carousel-highlight-slides) which was still referring to an independent `images` repo.  But the current file [Homepage.md](https://raw.githubusercontent.com/betterscientificsoftware/bssw.io/master/Site/Homepage.md) seems to show images that point into the bssw.io/images/ directory such as:

```
Slide4 R: images/raw/master/Blog_0721_PSIPhero_b.png
```

What is that `raw/master/` part for?  That makes no sense because I verified that that file `Blog_0721_PSIPhero_b.png` is only in `bssw.io/images/Blog_0721_PSIPhero_b.png` and not in the archived `images` repo.  I updated the documentation to match what I observed to be the current behavior but this needs to be fixed.  The `raw/master` part need to be taken out of that and it needs to be just:

```
Slide4 R: images/Blog_0721_PSIPhero_b.png
```

That file comes from the current branch of the bssw.io repo that is being used to generate the bssw.io site (or the preview.bssw.io site) and to leave `raw/master/` in there would seem to wrong and confusing unless I am totally confused about what is actually happening here.

**Comment by bartlettroscoe** [comment]
FYI: I searched the 'preview' branch and the 'master' branch and they both show three files that still point into the old images repo:

```
$ find Articles CuratedContent -name "*.md" -exec grep -l "github.com/betterscientificsoftware/images" {} \;
Articles/Blog/ApolloGuidanceComputerPart1.md
Articles/Blog/ApolloGuidanceComputerPart2.md
Articles/Blog/BlogArticleSkeletonA.md
```

I am not sure how these got missed in the conversion.  I will put in a PR that fixes these to point into the `images/` subdir.

**Comment by bartlettroscoe** [comment]
NOTE: With the merge of #1027 and #1028 and #On the 'master' and 'preview' branch esI ran:

```
$ find Articles CuratedContent -name "*.md" -exec grep -l "github.com/betterscientificsoftware/images" {} \;
```

and it comes up empty.

Therefore, I think all that is left is to deal with the handling of images for files under the `Site/` directory.  From @bernhold's comment in https://github.com/betterscientificsoftware/bssw.io/pull/1027#discussion_r692357424, it seems he and Parallactic are working on that.

**Comment by bernhold** [comment]
As of today, the production site is capable of using subdirectories in images.  I think we can declare this epic complete!
